### ISSUE TITLE
[Feature] Allow using nvtop as an library for other apps

### ISSUE THREAD

**Issue by pktiuk** [Syllo/nvtop#206]
I think this app does a great job as a tool for monitoring GPU usage for different vendors.     
But a CLI app is not always something people need.

I think, that with minor modifications this repository could be also used for building just a C (or C++) library `libnvtop`, which could be used by other developers to create other apps allowing monitoring GPUs.    
It could be used as a base for plugin for existing system monitors, and standalone GUI monitors, it could be used for monitoring servers used for machine learning with [cockpit](https://cockpit-project.org/) addon, etc.

**Comment by pktiuk** [comment]
@Syllo   
What do you think about this idea?

**Comment by Syllo** [comment]
While I agree that such library would be useful, I think that the current license would not be a good match for such library.

**Comment by pktiuk** [comment]
@Syllo   
What is current license of nvtop? And Why it is a wrong one?

I can only find a word `COPYING` in current license file.  https://github.com/Syllo/nvtop/blob/master/LICENSE

**Comment by pktiuk** [comment]
> What is wrong with current license? (GPL-3)

@Syllo (sorry for pinging, but I would really want to help with this issue)

**Comment by pktiuk** [comment]
> What is wrong with current license? (GPL-3)

@Syllo (sorry for pinging, but I would really like to help with this issue)

**Comment by topaxi** [comment]
I arrived here with a similar wish to use reading gpu metrics from a library and nvtop seems to support every gpu for my use-cases.

GPLv3 wouldn't be an issue as the project would be a personal toy project I could easily license as GPLv3 too. Alternatively it might not a big issue to have the license of the library as LGPLv3 while nvtop itself stays GPLv3, so non-GPL projects could link to the library (not necessary for me though :grin:).

**Comment by jobs-git** [comment]
> I arrived here with a similar wish to use reading gpu metrics from a library and nvtop seems to support every gpu for my use-cases.
> 
> GPLv3 wouldn't be an issue as the project would be a personal toy project I could easily license as GPLv3 too. Alternatively it might not a big issue to have the license of the library as LGPLv3 while nvtop itself stays GPLv3, so non-GPL projects could link to the library (not necessary for me though 😁).

Agreed. It would be the user's responsibility if it uses the library.

**Comment by pktiuk** [comment]
> Alternatively it might not a big issue to have the license of the library as LGPLv3 while nvtop itself stays GPLv3, so non-GPL projects could link to the library (not necessary for me though 😁).

AFAIK It would be an issue, because creating a library using GPL code from nvtop would require using GPL license in library. So a lot of code of nvtop should be relicensed to LGPL to make creating library doable.

**Comment by pktiuk** [comment]
@Syllo 

I think you should make a decision about this licensing issue. Because without it, we can't even think about creating a reasonably licensed library.

**Comment by Conan-Kudo** [comment]
> Of course there could be a GPL licensed library, but it would limit its usage only to GPL-licensed apps, because most of the other open-source licenses (like MIT) are not compatible with GPL.

This is not true. The combination of that is perfectly fine, the collective work just winds up being GPL. It doesn't make nvtop as a library any less viable.

**Comment by jobs-git** [comment]
> > It doesn't make nvtop as a library any less viable.
> 
> Well, I disagree. It makes libnvtop unusable for every already existing FOSS project with license different than GPL3 (GPL2 would be a bit better, because that license is a bit less restrictive and more FOSS projects could use it). Example apps which could use libnvtop, but are unable to because of license:
> 
> * [BTOP](https://github.com/aristocratos/btop) -  Apache-2.0 license
> * [Cockpit](https://cockpit-project.org/) - LGPL2.1
> * [Gnome System monitor](https://gitlab.gnome.org/GNOME/gnome-system-monitor) GPL2 (incompatible with GPL3)
> 
> Of course, there are still GPL3 projects which could use this library.

Why not create the PR first? Then update the license later? Currently its GPL3.
